{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#All packages we will preload for the code below.\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Seq import translate\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#%%%%%%%%%%%%%%%#\n",
    "### FUNCTIONS ###\n",
    "#%%%%%%%%%%%%%%%#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Function 1 ## Document the code provided to create a dictionary from the fasta file\n",
    "\n",
    "def get_sequences_from_file(fasta_fn): \n",
    "    #This is the start of the funciton, instead of calling all of this code later, we can simply call this line to run the code and insert the data in the ()\n",
    "    sequence_data_dict = {} \n",
    "    #This initializes a directory within the function to store information, this is a local dictionary and is not global, meaning it will get deleted unless stored globally\n",
    "    for record in SeqIO.parse(fasta_fn, \"fasta\"): \n",
    "    #This begins the for loop through the function, for each record within the \"fasta\" data file, do the following steps after this line\n",
    "        description = record.description.split() \n",
    "        #This takes the total description of the specimen from the fasta file and splits it into sections\n",
    "        species_name = description[1] + \" \" + description[2] \n",
    "        #This returns the 2nd object a space and the 3rd object as the dictionary key for each specimen\n",
    "        sequence_data_dict[species_name] = record.seq \n",
    "        #This populates the value for the key assigned above as the sequence record from the fasta file\n",
    "    return(sequence_data_dict) \n",
    "    #This returns the dictionary and finalizes the function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Function 2 ## Finish the psuedo-code for the translate_function to translate DNA sequence to Amino Acid sequence\n",
    "\n",
    "#Obtained from Stack Overflow at website https://stackoverflow.com/questions/19521905/translation-dna-to-protein\n",
    "#Modified to run within the confines of the data we are working on.\n",
    "#I found this code using google, but it works perfectly for what we are trying to achieve.\n",
    "#I will define the additional packages and tools utilized to turn the DNA sequence string into an Amino Acid Sequence String.\n",
    "\n",
    "from itertools import takewhile\n",
    "#The definition of itertools.takewhile from the python.org website \"Make an iterator that returns elements fromt the iterable as long as the preciate is true\"\n",
    "\n",
    "def translate_function(sequence, mito_table = CodonTable.unambiguous_dna_by_name[\"Vertebrate Mitochondrial\"], stop_codons = ('TAA', 'TAG', 'AGA', 'AGG')):\n",
    "    #This is the start of the function, sequence will be replaced by the data, within the function we initialize and define the codon table, define what are stop codons \n",
    "    start = sequence.find('ATG') \n",
    "    #This should find the start of the protein sequence, for our data all of our records start with ATG, but for other examples may be more applicable\n",
    "    trimmed_sequence = sequence[start:]\n",
    "    #This will trim the sequence to where we will begin the amino acid sequence string\n",
    "    codons = [trimmed_sequence[i:i+3] for i in range(0, len(trimmed_sequence), 3)]\n",
    "    #This will separate out the codons within the DNA sequence to groups of 3\n",
    "    coding_sequence = takewhile(lambda x: x not in stop_codons and len(x) == 3, codons)\n",
    "    #Here's where we will utilize itertools.takewhile, as long as the codon is not a stop codon it will be added to the amino acid sequence string\n",
    "    aa_seq_string = ''.join([mito_table.forward_table[codon] for codon in coding_sequence])\n",
    "    #This populates the sequence string by calling the appropriate amino acid from the mito_table and adding to the string\n",
    "    return \"{0}_\".format(aa_seq_string)\n",
    "    #This will return the string for the defined specimen put in the function()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cytb_seqs = get_sequences_from_file(\"penguins_cytb.fasta\")\n",
    "#Taken from the bottom, we will populate a global identity for the cytb_seqs dicitonary we created from the function above\n",
    "#Also a good test that the function works and doesn't spit an error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Aptenodytes forsteri': Seq('ATGGCCCCAAATCTCCGAAAATCCCATCCCCTCCTAAAAATAATTAATAACTCC...TAA', SingleLetterAlphabet()),\n",
       " 'Aptenodytes patagonicus': Seq('ATGGCCCCAAACCTCCGAAAATCCCATCCTCTCCTAAAAATAATTAATAACTCC...TAA', SingleLetterAlphabet()),\n",
       " 'Eudyptes chrysocome': Seq('ATGGCCCCCAACCTCCGAAAATCCCACCCCCTCCTAAAAACAATCAATAACTCC...TAA', SingleLetterAlphabet()),\n",
       " 'Eudyptes chrysolophus': Seq('ATGGCCCCCAACCTCCGAAAATCCCACCCCCTCCTAAAAACAATCAATAACTCC...TAA', SingleLetterAlphabet()),\n",
       " 'Eudyptes sclateri': Seq('ATGGCCCCCAACCTCCGAAAATCCCACCCCCTCCTAAAAACAATCAATAACTCC...TAA', SingleLetterAlphabet()),\n",
       " 'Eudyptula minor': Seq('ATGGCCCCCAACCTCCGAAAATCTCACCCCCTCCTAAAAATAATCAACAACTCT...TAA', SingleLetterAlphabet()),\n",
       " 'Pygoscelis adeliae': Seq('ATGGCCCCCAACCTCCGAAAATCCCACCCTCTCCTAAAAATAATTAACAACTCC...TAA', SingleLetterAlphabet()),\n",
       " 'Pygoscelis antarctica': Seq('ATGGCCCCCAACCTCCGAAAATCCCACCCTCTCCTAAAAATAATCAACAACTCC...TAG', SingleLetterAlphabet()),\n",
       " 'Pygoscelis papua': Seq('ATGGCCCCCAACCTTCGAAAATCCCACCCTCTCCTAAAAATAATCAACAAATCC...TAG', SingleLetterAlphabet()),\n",
       " 'Spheniscus demersus': Seq('ATGGCCCCCAACCTCCGAAAATCCCACCCTCTCCTAAAAACAATCAACAACTCC...TAA', SingleLetterAlphabet()),\n",
       " 'Spheniscus humboldti': Seq('ATGGCCCCCAACCTCCGAAAATCCCACCCTCTCCTAAAAACAATCAACAACTCC...TAA', SingleLetterAlphabet()),\n",
       " 'Spheniscus magellanicus': Seq('ATGGCCCCCAACCTCCGAAAATCCCACCCTCTCCTAAAAACAATCAACAACTCC...TAA', SingleLetterAlphabet())}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytb_seqs\n",
    "#A glance at the global dictionary to look at the keys and values\n",
    "#Again another good test for whether the function performed as we'd wanted it too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Stock\\Anaconda3\\lib\\site-packages\\Bio\\Seq.py:152: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  \"the new string hashing behaviour.\", BiopythonWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'MAPNLRKSHPLLKMINNSLIDLPTPSNISAWWNFGSLLGICLTTQILTGLLLAMHYTADTTLAFSSVAHTCRNVQYGWLIRNLHANGASFFFICIYLHIGRGFYYGSYLYKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWTWGGFSVDNPTLTRFFALHFLLPFMIAGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFALMLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLIPLLHKSKQRTMAFRPLSQLLFWALVANLIILTWVGSQPVEHPFIIIGQLASLTYFTTLLILFPIAGALENKMLNH_'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "translate_function(cytb_seqs['Aptenodytes forsteri'])\n",
    "#Testing the translate function for the first key in the cytb_seqs dictionary\n",
    "#It works but gives an interesting warning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MAPNLRKSHPLLKIINNSLIDLPTPSNISA**NFGSLLGICLTTQILTGLLLAI...NH*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Function 3 ## Write an easier function to obtain the Amino Acid sequence of a DNA sequence\n",
    "\n",
    "#There is a much easier way to translate a DNA sequence to Amino Acids that has already been written within the Biopython package\n",
    "#Taken from biopython.org tutorial, the translate function does almost exactly what the previous function we just created does\n",
    "#The minor difference is that in the translate function within biopython, the stop codons are still in the amino acid sequence but are designated with *\n",
    "#We will address that within our next function\n",
    "\n",
    "translate(cytb_seqs['Aptenodytes forsteri'])\n",
    "#For this example, I have again done the first key in the cytb_seqs dictinonary, which returns a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function 4 ## Calculate Molecular Weight of the Amino Acid Sequence\n",
    "\n",
    "#Pull the ProtParam module and import Protein Analysis for the next portion\n",
    "#http://biopython.org/wiki/ProtParam\n",
    "\n",
    "def molecular_weight(aa_seq):\n",
    "    \n",
    "    str_seq = ''.join(aa_seq)\n",
    "    #Converting the list to a string\n",
    "    str_seq = str_seq.replace(\"*\",\"\")\n",
    "    #Eliminating the stop codon positions\n",
    "    analysed_seq = ProteinAnalysis(str_seq)\n",
    "    #Uses the package to analyze the amino acid string\n",
    "    print(analysed_seq.molecular_weight())\n",
    "    #Print the calculated molecular weight of the string\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40471.2343\n"
     ]
    }
   ],
   "source": [
    "Apt_for = translate(cytb_seqs['Aptenodytes forsteri'])\n",
    "#Develop the list that we wish to test, again using the first key value of the dictionary we created\n",
    "\n",
    "molecular_weight(Apt_for)\n",
    "#Test the function, which seems to work, supplying the molecular weight of the aa_seq "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Function 5 ## Calculate the GC-content of a DNA sequence\n",
    "\n",
    "def GC_prop(sequence):\n",
    "    \n",
    "    g = sequence.count(\"G\")\n",
    "    #Count the number of G in the sequence\n",
    "    c = sequence.count(\"C\")\n",
    "    #Count the number of C in the sequence\n",
    "    l = len(sequence)\n",
    "    #Count the total length of the sequence\n",
    "    gc_prop = (g + c) / l\n",
    "    #Add counts of G and C and divide by total count to obtain proportion of GC Content\n",
    "    print(gc_prop)\n",
    "    #Print the proportion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.48381452318460194\n"
     ]
    }
   ],
   "source": [
    "GC_content(cytb_seqs['Aptenodytes forsteri'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#%%%%%%%%%%%%%%#\n",
    "###   MAIN   ###\n",
    "#%%%%%%%%%%%%%%#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cytb_seqs = get_sequences_from_file(\"penguins_cytb.fasta\") \n",
    "\n",
    "penguins_df = pd.read_csv(\"penguins_mass.csv\") # Includes only data for body mass \n",
    "species_list = list(penguins_df.species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>mass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aptenodytes forsteri</td>\n",
       "      <td>28.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aptenodytes patagonicus</td>\n",
       "      <td>13.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eudyptes chrysocome</td>\n",
       "      <td>2.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Eudyptes chrysolophus</td>\n",
       "      <td>4.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eudyptes sclateri</td>\n",
       "      <td>4.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Eudyptula minor</td>\n",
       "      <td>1.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pygoscelis adeliae</td>\n",
       "      <td>4.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pygoscelis antarctica</td>\n",
       "      <td>4.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Pygoscelis papua</td>\n",
       "      <td>6.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Spheniscus demersus</td>\n",
       "      <td>3.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Spheniscus humboldti</td>\n",
       "      <td>4.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Spheniscus magellanicus</td>\n",
       "      <td>3.40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    species   mass\n",
       "0      Aptenodytes forsteri  28.00\n",
       "1   Aptenodytes patagonicus  13.40\n",
       "2       Eudyptes chrysocome   2.80\n",
       "3     Eudyptes chrysolophus   4.50\n",
       "4         Eudyptes sclateri   4.25\n",
       "5           Eudyptula minor   1.60\n",
       "6        Pygoscelis adeliae   4.60\n",
       "7     Pygoscelis antarctica   4.10\n",
       "8          Pygoscelis papua   6.10\n",
       "9       Spheniscus demersus   3.20\n",
       "10     Spheniscus humboldti   4.75\n",
       "11  Spheniscus magellanicus   3.40"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "penguins_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 6 ## Add columns for \"molecular weight\" and \"GC content\" to the penguins_df\n",
    "\n",
    "#We have already imported numpy as np above\n",
    "#We will add each one separately using numpy and assigning the value for each as NaN to be filled in later\n",
    "\n",
    "penguins_df[\"molecular weight\"] = np.nan\n",
    "penguins_df[\"GC content\"] = np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40471.2343\n",
      "0.48381452318460194\n",
      "40743.510399999985\n",
      "0.4969378827646544\n",
      "40709.49419999999\n",
      "0.5118110236220472\n",
      "40865.67809999998\n",
      "0.5091863517060368\n",
      "40709.49419999999\n",
      "0.5083114610673666\n",
      "40335.1015\n",
      "0.4925634295713036\n",
      "40674.49449999999\n",
      "0.49081364829396323\n",
      "40193.8878\n",
      "0.4776902887139108\n",
      "40403.25979999999\n",
      "0.4715660542432196\n",
      "40238.9329\n",
      "0.48293963254593175\n",
      "40206.93589999999\n",
      "0.4925634295713036\n",
      "40266.986\n",
      "0.4820647419072616\n"
     ]
    }
   ],
   "source": [
    "for key, value in cytb_seqs.items():\n",
    "    \n",
    "    aa_seq = translate(value) \n",
    "    #Here we are using function #3 from above, the more simple approach\n",
    "    \n",
    "    mol_weight = molecular_weight(aa_seq)\n",
    "    #Here we will use function #4 from above\n",
    "    \n",
    "    GC_content = GC_prop(value)\n",
    "    #Here we will use function #5 from above\n",
    "    \n",
    "    #row = list(cytb_seqs.keys()).index(key)\n",
    "    \n",
    "    #penguins_df[row ,'molecular weight'] = mol_weight\n",
    "    #penguins_df[row ,'GC Content'] = GC_content\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>mass</th>\n",
       "      <th>molecular weight</th>\n",
       "      <th>GC content</th>\n",
       "      <th>GC Content</th>\n",
       "      <th>(0, molecular weight)</th>\n",
       "      <th>(0, GC Content)</th>\n",
       "      <th>(1, molecular weight)</th>\n",
       "      <th>(1, GC Content)</th>\n",
       "      <th>(2, molecular weight)</th>\n",
       "      <th>...</th>\n",
       "      <th>(7, molecular weight)</th>\n",
       "      <th>(7, GC Content)</th>\n",
       "      <th>(8, molecular weight)</th>\n",
       "      <th>(8, GC Content)</th>\n",
       "      <th>(9, molecular weight)</th>\n",
       "      <th>(9, GC Content)</th>\n",
       "      <th>(10, molecular weight)</th>\n",
       "      <th>(10, GC Content)</th>\n",
       "      <th>(11, molecular weight)</th>\n",
       "      <th>(11, GC Content)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aptenodytes forsteri</td>\n",
       "      <td>28.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aptenodytes patagonicus</td>\n",
       "      <td>13.40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eudyptes chrysocome</td>\n",
       "      <td>2.80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Eudyptes chrysolophus</td>\n",
       "      <td>4.50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eudyptes sclateri</td>\n",
       "      <td>4.25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Eudyptula minor</td>\n",
       "      <td>1.60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pygoscelis adeliae</td>\n",
       "      <td>4.60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pygoscelis antarctica</td>\n",
       "      <td>4.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Pygoscelis papua</td>\n",
       "      <td>6.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Spheniscus demersus</td>\n",
       "      <td>3.20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Spheniscus humboldti</td>\n",
       "      <td>4.75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Spheniscus magellanicus</td>\n",
       "      <td>3.40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    species   mass  molecular weight  GC content  GC Content  \\\n",
       "0      Aptenodytes forsteri  28.00               NaN         NaN         NaN   \n",
       "1   Aptenodytes patagonicus  13.40               NaN         NaN         NaN   \n",
       "2       Eudyptes chrysocome   2.80               NaN         NaN         NaN   \n",
       "3     Eudyptes chrysolophus   4.50               NaN         NaN         NaN   \n",
       "4         Eudyptes sclateri   4.25               NaN         NaN         NaN   \n",
       "5           Eudyptula minor   1.60               NaN         NaN         NaN   \n",
       "6        Pygoscelis adeliae   4.60               NaN         NaN         NaN   \n",
       "7     Pygoscelis antarctica   4.10               NaN         NaN         NaN   \n",
       "8          Pygoscelis papua   6.10               NaN         NaN         NaN   \n",
       "9       Spheniscus demersus   3.20               NaN         NaN         NaN   \n",
       "10     Spheniscus humboldti   4.75               NaN         NaN         NaN   \n",
       "11  Spheniscus magellanicus   3.40               NaN         NaN         NaN   \n",
       "\n",
       "   (0, molecular weight) (0, GC Content) (1, molecular weight)  \\\n",
       "0                   None            None                  None   \n",
       "1                   None            None                  None   \n",
       "2                   None            None                  None   \n",
       "3                   None            None                  None   \n",
       "4                   None            None                  None   \n",
       "5                   None            None                  None   \n",
       "6                   None            None                  None   \n",
       "7                   None            None                  None   \n",
       "8                   None            None                  None   \n",
       "9                   None            None                  None   \n",
       "10                  None            None                  None   \n",
       "11                  None            None                  None   \n",
       "\n",
       "   (1, GC Content) (2, molecular weight)       ...         \\\n",
       "0             None                  None       ...          \n",
       "1             None                  None       ...          \n",
       "2             None                  None       ...          \n",
       "3             None                  None       ...          \n",
       "4             None                  None       ...          \n",
       "5             None                  None       ...          \n",
       "6             None                  None       ...          \n",
       "7             None                  None       ...          \n",
       "8             None                  None       ...          \n",
       "9             None                  None       ...          \n",
       "10            None                  None       ...          \n",
       "11            None                  None       ...          \n",
       "\n",
       "   (7, molecular weight) (7, GC Content) (8, molecular weight)  \\\n",
       "0                   None            None                  None   \n",
       "1                   None            None                  None   \n",
       "2                   None            None                  None   \n",
       "3                   None            None                  None   \n",
       "4                   None            None                  None   \n",
       "5                   None            None                  None   \n",
       "6                   None            None                  None   \n",
       "7                   None            None                  None   \n",
       "8                   None            None                  None   \n",
       "9                   None            None                  None   \n",
       "10                  None            None                  None   \n",
       "11                  None            None                  None   \n",
       "\n",
       "   (8, GC Content) (9, molecular weight) (9, GC Content)  \\\n",
       "0             None                  None            None   \n",
       "1             None                  None            None   \n",
       "2             None                  None            None   \n",
       "3             None                  None            None   \n",
       "4             None                  None            None   \n",
       "5             None                  None            None   \n",
       "6             None                  None            None   \n",
       "7             None                  None            None   \n",
       "8             None                  None            None   \n",
       "9             None                  None            None   \n",
       "10            None                  None            None   \n",
       "11            None                  None            None   \n",
       "\n",
       "   (10, molecular weight) (10, GC Content) (11, molecular weight)  \\\n",
       "0                    None             None                   None   \n",
       "1                    None             None                   None   \n",
       "2                    None             None                   None   \n",
       "3                    None             None                   None   \n",
       "4                    None             None                   None   \n",
       "5                    None             None                   None   \n",
       "6                    None             None                   None   \n",
       "7                    None             None                   None   \n",
       "8                    None             None                   None   \n",
       "9                    None             None                   None   \n",
       "10                   None             None                   None   \n",
       "11                   None             None                   None   \n",
       "\n",
       "   (11, GC Content)  \n",
       "0              None  \n",
       "1              None  \n",
       "2              None  \n",
       "3              None  \n",
       "4              None  \n",
       "5              None  \n",
       "6              None  \n",
       "7              None  \n",
       "8              None  \n",
       "9              None  \n",
       "10             None  \n",
       "11             None  \n",
       "\n",
       "[12 rows x 29 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "penguins_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(cytb_seqs.keys()).index(\"Aptenodytes forsteri\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
